storage/store_pool_test.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Bram Gruneir (bram+code@cockroachlabs.com)
+
+package storage
+
+import (
+	"errors"
+	"reflect"
+	"sort"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/base"
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/rpc"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/stop"
+)
+
+var uniqueStore = []*proto.StoreDescriptor{
+	{
+		StoreID: 2,
+		Attrs:   proto.Attributes{Attrs: []string{"ssd"}},
+		Node: proto.NodeDescriptor{
+			NodeID: 2,
+			Attrs:  proto.Attributes{Attrs: []string{"a"}},
+		},
+		Capacity: proto.StoreCapacity{
+			Capacity:  100,
+			Available: 200,
+		},
+	},
+}
+
+// createTestStorePool creates a stopper, gossip and storePool for use in
+// tests. Stopper must be stopped by the caller.
+func createTestStorePool() (*stop.Stopper, *gossip.Gossip, *StorePool) {
+	stopper := stop.NewStopper()
+	rpcContext := rpc.NewContext(&base.Context{}, hlc.NewClock(hlc.UnixNano), stopper)
+	g := gossip.New(rpcContext, gossip.TestInterval, gossip.TestBootstrap)
+	storePool := NewStorePool(g, testTimeUntilStoreDead, stopper)
+	return stopper, g, storePool
+}
+
+// TestStorePoolGossipUpdate ensures that the gossip callback in StorePool
+// correctly updates a store's details.
+func TestStorePoolGossipUpdate(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	stopper, g, sp := createTestStorePool()
+	defer stopper.Stop()
+	sg := newStoreGossiper(g)
+
+	sp.mu.RLock()
+	if _, ok := sp.stores[2]; ok {
+		t.Fatalf("store 2 is already in the pool's store list")
+	}
+	sp.mu.RUnlock()
+
+	sg.gossipStores(uniqueStore, t)
+
+	sp.mu.RLock()
+	if _, ok := sp.stores[2]; !ok {
+		t.Fatalf("store 2 isn't in the pool's store list")
+	}
+	if e, a := 1, sp.queue.Len(); e > a {
+		t.Fatalf("wrong number of stores in the queue expected at least:%d actual:%d", e, a)
+	}
+	sp.mu.RUnlock()
+}
+
+// waitUntilDead will block until the specified store is marked as dead.
+func waitUntilDead(t *testing.T, sp *StorePool, storeID proto.StoreID) {
+	util.SucceedsWithin(t, 3*testTimeUntilStoreDead, func() error {
```

Comment:
Look at description for isTrueWithin:
`// This method is deprecated; use SucceedsWithin instead.`